### PR TITLE
Miner Style Points 2: Style on Everyone

### PR DESCRIPTION
**Pull request by Zonespace27**
## About The Pull Request
A re-open of https://github.com/tgstation/tgstation/pull/66326 with Fikou's permission

Adds the style meter, it can be bought from the mining vendor for 1500 points, it is an attachment to your glasses.
The style meter creates a display on your hud, with your recent actions, like attacking enemies, killing them, mining ore etc. Actions like spinning or flipping increase your score multiplier, making you get more points.
Your style meter affects how much ore you get from mining rocks. By default with the meter, you get 20% less ore, but at the highest, you can get 1.2x the ore from mining. In addition, on B-tier or above, you can "hotswap" items, by attacking an item in your backpack with one in your hand (should it fit and all that). Also features a leaderboard for highest style point count!

New streamable: https://streamable.com/eewi6l

The following are sources of points:

- Killing things
- Killing big things
- Killing small things
- Punching things
- Melee'ing things
- Mining rocks and ores
- Having matrix traps detonate
- Hit, defuse, and detonate gibtonite
- Detonate crusher marks
- Scan geysers
- Parry projectiles (others or your own)

Oh, right. While wearing the style meter, you're able to parry any lavaland-based projectile by clicking on it or the tile it is on, which reflects it back in a 7 degree arc, making it 20% faster and 15% more damaging. Usually not very easy.

Maybe-plan in the future for some syndicate variant of this (with bullet parrying and appropriate style sources, etc.), but not for this PR

Thanks to Arcane, multitooling the style meter will make it play some sounds on rank-up.

image

https://streamable.com/nheaky

Parrying in action

## Why It's Good For The Game
Makes miners bring more ore in a fun way.

## Changelog
:cl: Fikou, Zonespace, Arcane for voicing
add: The mining vendor now has a style meter. This meter gauges your style points and uses them to improve your ore yield.
/:cl:

### CODE REVIEW

**Comment by LePiromano** [comment]
SO FUCKING GOOD

**Comment by dawsonkeyes** [comment]
### I'M GOING TO ULTRAKILL YOU

**Comment by Livrah** [comment]
Time to parry fireball back to Wizard and my life will be complete.

**Comment by Sneeker134** [comment]
It looks like parrying has a cooldown **after** you successfully parry, but unless I'm an idiot (very possible) it doesn't seem to have one for attempting a parry. Wouldn't parrying a lot of stuff become rather trivial by using an autoclicker, or even just spamming click really fast? Having a cooldown for you failing a parry would fix this, but that might introduce some other issues.

File: code/datums/components/style/style_meter.dm [hunk 1]
```diff
@@ -8,14 +8,20 @@
...
+	. = ..()
+	meter_appearance = mutable_appearance(icon, icon_state)
 
 /obj/item/style_meter/afterattack(atom/movable/attacked_atom, mob/user, proximity_flag, click_parameters)
 	. = ..()
 	if(!istype(attacked_atom, /obj/item/clothing/glasses))
 		return
 
 	forceMove(attacked_atom)
-	attacked_atom.vis_contents += src
```
Comment (Fikou) [inline code/datums/components/style/style_meter.dm]: why

Comment (Zonespace27) [reply code/datums/components/style/style_meter.dm]: It didn't work at all, the meter didn't appear
image

new method works though

**Comment by GuillaumePrata** [comment]
Having something as powerful as reflecting projectiles, regardless of how hard it would be to pull out, on a mostly invisible eyewear item sounds quite unfun to deal with.

It is the kind of feature that would make more sense on the Cursed or Ninja Katana.

**Comment by LemonInTheDark** [comment]
overlaps with the changeling ui element, need to figure something out for that

**Comment by Iamgoofball** [comment]
@LemonInTheDark IC issue, the flashy heads up display makes it hard for them to focus inward on their levels

**Comment by DaydreamIQ** [comment]
The deflect mechanic should either not work on station or shouldn't exist at all cause that's REALLY powerful for a miner item they can just buy without any trouble

**Comment by necromanceranne** [comment]
Just a question, but if you click a tile with multiple projectiles on it, whille it deflect all the projectiles on that tile, or just one?

**Comment by moocowswag** [comment]
gloves of north star to stack up style meter when?

**Comment by VladinXXV** [comment]
In regards to the projectile reflect concerns, I think the best way to deal with that is a whitelist of lavaland-oriented projectiles. There aren't that many. It'd keep the reflect useful for miners (even if the fauna make it on-station) without being discount sleeping carp. Obviously fauna projectiles would go on there. I personally could entertain including plasmacutter beams and PKA blasts, but that might not roll so well with others.

**Comment by Zonespace27** [comment]
Talked w/ maints, doing a few things when I have the time

**Comment by Toparai** [comment]
Do it, machine.
Run at the nukie at full speed with your cool glasses.

**Comment by ilikesaying** [comment]
>You have to punch the projectile with an empty hand

that seems fair and balanced, gonna have to keep a hand empty in order to be able to parry, so no dual wielding or wielding, or you're gonna have to drop/stash/micromanage your weapon/item.

what if the clicking empty tile part was locked behind an expensive upgrade?

**Comment by Zonespace27** [comment]
> what if the clicking empty tile part was locked behind an expensive upgrade?

Because it is nigh-impossible to directly click on anything but the slowest of projectiles

**Comment by dopamiin0** [comment]
in a future where having spin and flip binds is meta...

**Comment by MrrFiish** [comment]
OKAY GUYS BUT HEAR ME OUT
This is cool and all, but what if we ALSO added a SYNDICATE style meter for traitors and/or nukies, I have NO fucking clue what it would do differently but it would be so fucking cool if this existed. That is all.

**Comment by Zonespace27** [comment]
@Fikou Changes made

- Added hotswapping on b-tier and above
- Parrying is mining-projectile exclusive
- Reduced max ore gain to +20%

**Comment by CoiledLamb** [comment]
i want this to be merged so bad, hope these changes are enough

File: code/datums/components/style/style.dm [hunk 2]
```diff
@@ -0,0 +1,445 @@
...
+	change_points(-30, use_multiplier = FALSE)
+
+/datum/component/style/proc/on_death(datum/source, mob/living/died, gibbed)
+	SIGNAL_HANDLER
+
+	var/mob/mob_parent = parent
+	if(died == parent)
+		change_points(-500, use_multiplier = FALSE)
+		return
+	else if(faction_check(mob_parent.faction, died.faction) || (died.z != mob_parent.z) || !(died in view(mob_parent.client?.view, get_turf(mob_parent))))
```
Comment (Fikou) [inline code/datums/components/style/style.dm:414]: what if there is no client

File: code/datums/components/style/style.dm [hunk 3]
```diff
@@ -0,0 +1,445 @@
...
+	UnregisterSignal(parent, COMSIG_MOB_APPLY_DAMAGE)
+	UnregisterSignal(parent, list(COMSIG_MOB_EMOTED("flip"), COMSIG_MOB_EMOTED("spin")))
+	UnregisterSignal(parent, list(COMSIG_MOB_ITEM_ATTACK, COMSIG_HUMAN_MELEE_UNARMED_ATTACK))
+	UnregisterSignal(SSdcs, COMSIG_GLOB_MOB_DEATH)
+	UnregisterSignal(parent, COMSIG_LIVING_RESONATOR_BURST)
+	UnregisterSignal(parent, COMSIG_LIVING_PROJECTILE_PARRIED)
+	UnregisterSignal(parent, COMSIG_LIVING_DEFUSED_GIBTONITE)
+	UnregisterSignal(parent, COMSIG_LIVING_DISCOVERED_GEYSER)
+
+	qdel(projectile_parry?.resolve())
```
Comment (Fikou) [inline code/datums/components/style/style.dm:100]: use RemoveComponentSource

File: code/datums/components/style/style.dm [hunk 4]
```diff
@@ -0,0 +1,445 @@
...
+	RegisterSignal(parent, COMSIG_HUMAN_MELEE_UNARMED_ATTACK, PROC_REF(on_punch))
+	RegisterSignal(SSdcs, COMSIG_GLOB_MOB_DEATH, PROC_REF(on_death))
+	RegisterSignal(parent, COMSIG_LIVING_RESONATOR_BURST, PROC_REF(on_resonator_burst))
+	RegisterSignal(parent, COMSIG_LIVING_PROJECTILE_PARRIED, PROC_REF(on_projectile_parry))
+	RegisterSignal(parent, COMSIG_LIVING_DEFUSED_GIBTONITE, PROC_REF(on_gibtonite_defuse))
+	RegisterSignal(parent, COMSIG_LIVING_CRUSHER_DETONATE, PROC_REF(on_crusher_detonate))
+	RegisterSignal(parent, COMSIG_LIVING_DISCOVERED_GEYSER, PROC_REF(on_geyser_discover))
+
+	projectile_parry = WEAKREF(parent.AddComponentFrom(\
+		src,\
```
Comment (Fikou) [inline code/datums/components/style/style.dm:78]: would prefer REF(src)

File: code/datums/components/style/style.dm [hunk 5]
```diff
@@ -0,0 +1,445 @@
...
+	INVOKE_ASYNC(source, TYPE_PROC_REF(/mob/living, put_in_hands), target)
+	source.visible_message(span_notice("[source] quickly swaps [weapon] out with [target]!"), span_notice("You quickly swap [weapon] with [target]."))
+
+
+
+// Point givers
+/datum/component/style/proc/on_punch(mob/living/carbon/human/punching_person, atom/attacked_atom, proximity)
+	SIGNAL_HANDLER
+
+	if(!proximity || !punching_person.combat_mode || !istype(attacked_atom, /mob/living/simple_animal/hostile))
```
Comment (Fikou) [inline code/datums/components/style/style.dm:303]: what about basic mobs

File: code/datums/components/style/style.dm [hunk 6]
```diff
@@ -0,0 +1,445 @@
...
+
+	add_action(ACTION_DISRESPECT, 60 * (ismegafauna(disrespected) ? 2 : 1))
+
+/datum/component/style/proc/on_attack(mob/living/attacking_person, mob/living/attacked_mob)
+	SIGNAL_HANDLER
+
+	if(attacked_mob.stat)
+		return
+
+	var/mob/living/simple_animal/hostile/attacked_hostile = attacked_mob
```
Comment (Fikou) [inline code/datums/components/style/style.dm:318]: how do you know its a hostile?

Comment (Fikou) [reply code/datums/components/style/style.dm:318]: why do you even need it to be a hostile

File: code/datums/components/parry.dm [hunk 7]
```diff
@@ -0,0 +1,39 @@
+/// Add to a living mob to allow them to "parry" projectiles by clicking on their tile, sending them back at the firer.
+/datum/component/projectile_parry
+	dupe_mode = COMPONENT_DUPE_SOURCES
```
Comment (Fikou) [inline code/datums/components/parry.dm:3]: after looking at dupe sources its... not the way i envisioned it much. this'll have to be a normal component for now

Comment (Zonespace27) [reply code/datums/components/parry.dm:3]: sure thing chief

File: code/datums/components/style/style.dm (hunk 4 above)
Comment (Zonespace27) [reply code/datums/components/style/style.dm:78]: i just converted it into a full ref

File: code/datums/components/style/style.dm (hunk 2 above)
Comment (Zonespace27) [reply code/datums/components/style/style.dm:414]: the ?

File: code/datums/components/style/style.dm (hunk 5 above)
Comment (Zonespace27) [reply code/datums/components/style/style.dm:303]: converted to living

File: code/datums/components/style/style.dm (hunk 6 above)
Comment (Zonespace27) [reply code/datums/components/style/style.dm:318]: it doesn't

File: code/datums/components/style/style.dm [hunk 8]
```diff
@@ -0,0 +1,496 @@
...
+	RegisterSignal(parent, COMSIG_HUMAN_MELEE_UNARMED_ATTACK, PROC_REF(on_punch))
+	RegisterSignal(SSdcs, COMSIG_GLOB_MOB_DEATH, PROC_REF(on_death))
+	RegisterSignal(parent, COMSIG_LIVING_RESONATOR_BURST, PROC_REF(on_resonator_burst))
+	RegisterSignal(parent, COMSIG_LIVING_PROJECTILE_PARRIED, PROC_REF(on_projectile_parry))
+	RegisterSignal(parent, COMSIG_LIVING_DEFUSED_GIBTONITE, PROC_REF(on_gibtonite_defuse))
+	RegisterSignal(parent, COMSIG_LIVING_CRUSHER_DETONATE, PROC_REF(on_crusher_detonate))
+	RegisterSignal(parent, COMSIG_LIVING_DISCOVERED_GEYSER, PROC_REF(on_geyser_discover))
+
+	projectile_parry = AddComponent(\
+		src,\
```
Comment (Fikou) [inline code/datums/components/style/style.dm:117]: er. You didn't get what i meant. You passed in src as the source argument for the sourced component part of parrying. i think it should be REF instead. this variable should still be a weakref.

File: code/datums/components/style/style.dm [hunk 9]
```diff
@@ -0,0 +1,496 @@
...
+/datum/component/style/proc/remove_action(action_id)
+	actions -= action_id
+	update_screen()
+
+/datum/component/style/proc/change_points(amount, use_multiplier = TRUE)
+	if(!amount)
+		return
+
+	var/modified_amount = amount * (amount > 0 ? 1 - 0.1 * rank : 1) * (use_multiplier ? point_multiplier : 1)
+	style_points = clamp(style_points + modified_amount, -1, 499)
```
Comment (Fikou) [inline code/datums/components/style/style.dm:185]: you added a high score but its clamped to 499

File: code/datums/components/style/style.dm (hunk 8 above)
Comment (Zonespace27) [reply code/datums/components/style/style.dm:117]: oh yeah forgot to remove src, whoops

File: code/datums/components/style/style.dm (hunk 9 above)
Comment (Zonespace27) [reply code/datums/components/style/style.dm]: fixed

File: code/datums/components/style/style.dm [hunk 10]
```diff
@@ -0,0 +1,495 @@
...
+	RegisterSignal(parent, COMSIG_MOB_ITEM_ATTACK, PROC_REF(on_attack))
+	RegisterSignal(parent, COMSIG_HUMAN_MELEE_UNARMED_ATTACK, PROC_REF(on_punch))
+	RegisterSignal(SSdcs, COMSIG_GLOB_MOB_DEATH, PROC_REF(on_death))
+	RegisterSignal(parent, COMSIG_LIVING_RESONATOR_BURST, PROC_REF(on_resonator_burst))
+	RegisterSignal(parent, COMSIG_LIVING_PROJECTILE_PARRIED, PROC_REF(on_projectile_parry))
+	RegisterSignal(parent, COMSIG_LIVING_DEFUSED_GIBTONITE, PROC_REF(on_gibtonite_defuse))
+	RegisterSignal(parent, COMSIG_LIVING_CRUSHER_DETONATE, PROC_REF(on_crusher_detonate))
+	RegisterSignal(parent, COMSIG_LIVING_DISCOVERED_GEYSER, PROC_REF(on_geyser_discover))
+
+	projectile_parry = AddComponent(\
```
Comment (Fikou) [inline code/datums/components/style/style.dm:116]: again, should be a weakref

Comment (LemonInTheDark) [reply code/datums/components/style/style.dm:116]: preferably wouldn't exist, but hhhh

**Comment by Fikou** [comment]
image

**Pull request merged**